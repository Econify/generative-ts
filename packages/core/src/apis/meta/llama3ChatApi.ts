import type { ModelApi, ModelRequestOptions } from "../../typeDefs";

import { Template } from "../../utils/template";

import type { FewShotRequestOptions } from "../_shared_interfaces/fewShot";

import { isLlamaResponse, LlamaResponse } from "./llama";

const llama3ChatMlTemplateSrc = `<|begin_of_text|><% if (typeof system !== 'undefined') { %><|start_header_id|>system<|end_header_id|>\\n\\n<%= system %><|eot_id|><% } %><% (typeof examplePairs !== 'undefined' ? examplePairs : []).forEach(pair => { %><|start_header_id|>user<|end_header_id|>\\n\\n<%= pair.user %><|eot_id|><|start_header_id|>assistant<|end_header_id|>\\n\\n<%= pair.assistant %><|eot_id|><% }) %><|start_header_id|>user<|end_header_id|>\\n\\n<%= prompt %><|eot_id|><|start_header_id|>assistant<|end_header_id|>`;

const templateSource = `{
  "prompt": "${llama3ChatMlTemplateSrc}"
  <% if (typeof temperature !== 'undefined') { %>, "temperature": <%= temperature %><% } %>
  <% if (typeof top_p !== 'undefined') { %>, "top_p": <%= top_p %><% } %>
  <% if (typeof max_gen_len !== 'undefined') { %>, "max_gen_len": <%= max_gen_len %><% } %>
}`;

export interface Llama3ChatOptions
  extends FewShotRequestOptions,
    ModelRequestOptions {
  temperature?: number;
  top_p?: number;
  max_gen_len?: number;
}

export const Llama3Template = new Template<Llama3ChatOptions>(templateSource);

export const Llama3ChatApi: ModelApi<Llama3ChatOptions, LlamaResponse> = {
  requestTemplate: Llama3Template,
  responseGuard: isLlamaResponse,
};
