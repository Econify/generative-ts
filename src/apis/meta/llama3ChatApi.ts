import type { ModelApi, ModelRequestOptions } from "../../typeDefs";

import { Template } from "../../utils/template";

import type { FewShotRequestOptions } from "../shared/fewShot";

import { isLlamaResponse, LlamaResponse } from "./llama";

const llama3ChatPrompt =
  "<|begin_of_text|>" +
  "{% if system %}<|start_header_id|>system<|end_header_id|>\\n\\n{{ system | safe }}<|eot_id|>{% endif %}" +
  "{% for pair in examplePairs %}<|start_header_id|>user<|end_header_id|>\\n\\n{{ pair.user | safe }}<|eot_id|><|start_header_id|>assistant<|end_header_id|>\\n\\n{{ pair.assistant | safe }}<|eot_id|>{% endfor %}" +
  "<|start_header_id|>user<|end_header_id|>\\n\\n{{ prompt | safe }}<|eot_id|><|start_header_id|>assistant<|end_header_id|>";

export const templateSource =
  "{" +
  `"prompt": "${llama3ChatPrompt}"` +
  '{% if temperature %}, "temperature": {{ temperature }}{% endif %}' +
  '{% if top_p %}, "top_p": {{ top_p }}{% endif %}' +
  '{% if max_gen_len %}, "max_gen_len": {{ max_gen_len }}{% endif %}' +
  "}";

export interface Llama3ChatOptions
  extends FewShotRequestOptions,
    ModelRequestOptions {
  temperature?: number;
  top_p?: number;
  max_gen_len?: number;
}

export const Llama3Template = new Template<Llama3ChatOptions>(templateSource);

export const Llama3ChatApi: ModelApi<Llama3ChatOptions, LlamaResponse> = {
  requestTemplate: Llama3Template,
  responseGuard: isLlamaResponse,
};
