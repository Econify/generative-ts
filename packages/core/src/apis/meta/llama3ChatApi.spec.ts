import { Llama3ChatApi, Llama3ChatOptions } from "./llama3ChatApi";

function render(context: Omit<Llama3ChatOptions, "modelId">) {
  try {
    const rendered = Llama3ChatApi.requestTemplate.render({
      modelId: "mock-model-id",
      ...context,
    });
    return JSON.parse(rendered);
  } catch (error) {
    throw new Error("Invalid JSON output");
  }
}

describe("Llama3ChatApi:", () => {
  /**
   * FewShotRequestOptions (prompt, examplePairs, system):
   */

  test("prompt", () => {
    const rendered = render({
      prompt: "mock-prompt",
    });

    expect(rendered).toMatchSnapshot();
  });

  test("prompt, system", () => {
    const rendered = render({
      prompt: "mock-prompt",
      system: "mock-system-text",
    });

    expect(rendered).toMatchSnapshot();
  });

  test("prompt, examplePairs", () => {
    const rendered = render({
      prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
        { user: "mock-user-msg-2", assistant: "mock-assistant-msg-2" },
      ],
    });

    expect(rendered).toMatchSnapshot();
  });

  test("prompt, system, examplePairs", () => {
    const rendered = render({
      prompt: "mock-prompt",
      system: "mock-system-text",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
        { user: "mock-user-msg-2", assistant: "mock-assistant-msg-2" },
      ],
    });

    expect(rendered).toMatchSnapshot();
  });

  /**
   * All options:
   */

  test("all options", () => {
    const rendered = render({
      prompt: "mock-prompt",
      system: "mock-system-text",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
        { user: "mock-user-msg-2", assistant: "mock-assistant-msg-2" },
      ],
      temperature: 0.7,
      top_p: 0.9,
      max_gen_len: 512,
    });

    expect(rendered).toMatchSnapshot();
  });
});
