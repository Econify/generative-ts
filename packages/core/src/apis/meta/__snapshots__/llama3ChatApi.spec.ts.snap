// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Llama3ChatApi: $prompt 1`] = `
{
  "prompt": "<|begin_of_text|><|start_header_id|>user<|end_header_id|>

mock-prompt<|eot_id|><|start_header_id|>assistant<|end_header_id|>",
}
`;

exports[`Llama3ChatApi: $prompt, examplePairs 1`] = `
{
  "prompt": "<|begin_of_text|><|start_header_id|>user<|end_header_id|>

mock-user-msg-1<|eot_id|><|start_header_id|>assistant<|end_header_id|>

mock-assistant-msg-1<|eot_id|><|start_header_id|>user<|end_header_id|>

mock-user-msg-2<|eot_id|><|start_header_id|>assistant<|end_header_id|>

mock-assistant-msg-2<|eot_id|><|start_header_id|>user<|end_header_id|>

mock-prompt<|eot_id|><|start_header_id|>assistant<|end_header_id|>",
}
`;

exports[`Llama3ChatApi: $prompt, system 1`] = `
{
  "prompt": "<|begin_of_text|><|start_header_id|>system<|end_header_id|>

mock-system-text<|eot_id|><|start_header_id|>user<|end_header_id|>

mock-prompt<|eot_id|><|start_header_id|>assistant<|end_header_id|>",
}
`;

exports[`Llama3ChatApi: $prompt, system, examplePairs 1`] = `
{
  "prompt": "<|begin_of_text|><|start_header_id|>system<|end_header_id|>

mock-system-text<|eot_id|><|start_header_id|>user<|end_header_id|>

mock-user-msg-1<|eot_id|><|start_header_id|>assistant<|end_header_id|>

mock-assistant-msg-1<|eot_id|><|start_header_id|>user<|end_header_id|>

mock-user-msg-2<|eot_id|><|start_header_id|>assistant<|end_header_id|>

mock-assistant-msg-2<|eot_id|><|start_header_id|>user<|end_header_id|>

mock-prompt<|eot_id|><|start_header_id|>assistant<|end_header_id|>",
}
`;

exports[`Llama3ChatApi: all options 1`] = `
{
  "max_gen_len": 512,
  "prompt": "<|begin_of_text|><|start_header_id|>system<|end_header_id|>

mock-system-text<|eot_id|><|start_header_id|>user<|end_header_id|>

mock-user-msg-1<|eot_id|><|start_header_id|>assistant<|end_header_id|>

mock-assistant-msg-1<|eot_id|><|start_header_id|>user<|end_header_id|>

mock-user-msg-2<|eot_id|><|start_header_id|>assistant<|end_header_id|>

mock-assistant-msg-2<|eot_id|><|start_header_id|>user<|end_header_id|>

mock-prompt<|eot_id|><|start_header_id|>assistant<|end_header_id|>",
  "temperature": 0.7,
  "top_p": 0.9,
}
`;
